sdk/core/System.ClientModel/src/Internal/Logging/LoggingHandler.cs

```diff
@@ -0,0 +1,319 @@
+        _logger = logger;
+        _sanitizer = sanitizer;
+        _logMessages = new ClientModelLogMessages(logger);
+        _useILogger = logger is not NullLogger;
+    }
+
+    /// <summary>
+    ///  Whether the given log level or event level is enabled, depending
+    ///  on whether this handler logs to ILogger or Event Source. Can be
+    ///  used to guard expensive operations.
+    /// </summary>
+    /// <param name="logLevel"></param>
+    /// <param name="eventLevel"></param>
+    public bool IsEnabled(LogLevel logLevel, EventLevel eventLevel)
+    {
+        return _useILogger ? _logger.IsEnabled(logLevel) : ClientModelEventSource.Log.IsEnabled(eventLevel, EventKeywords.None);
+    }
+
+    public void LogRequest(PipelineRequest request, string requestId, string? assemblyName)
+    {
+        if (_useILogger)
+        {
+            if (_logger.IsEnabled(LogLevel.Information))
+            {
+                _logMessages.Request(requestId, request.Method, _sanitizer.SanitizeUrl(request.Uri!.AbsoluteUri), FormatHeaders(request.Headers), assemblyName);
+            }
+        }
+        else
+        {
+            if (ClientModelEventSource.Log.IsEnabled(EventLevel.Informational, EventKeywords.None))
+            {
+                ClientModelEventSource.Log.Request(requestId, request.Method, _sanitizer.SanitizeUrl(request.Uri!.AbsoluteUri), FormatHeaders(request.Headers), assemblyName);
+            }
+        }
+    }
+
+    public void LogRequestContent(string requestId, byte[] content, Encoding? textEncoding)
+    {
+        if (_useILogger)
+        {
+            if (_logger.IsEnabled(LogLevel.Debug))
+            {
+                if (textEncoding != null)
+                {
+                    _logMessages.RequestContentText(requestId, textEncoding.GetString(content));
+                }
+                else
+                {
+                    _logMessages.RequestContent(requestId, content);
+                }
+            }
+        }
+        else
+        {
+            if (ClientModelEventSource.Log.IsEnabled(EventLevel.Verbose, EventKeywords.None))
+            {
+                if (textEncoding != null)
+                {
+                    ClientModelEventSource.Log.RequestContentText(requestId, textEncoding.GetString(content));
+                }
+                else
+                {
+                    ClientModelEventSource.Log.RequestContent(requestId, content);
+                }
+            }
+        }
+    }
+
+    public void LogResponse(string requestId, PipelineResponse response, double seconds)
+    {
+        if (_useILogger)
+        {
+            if (_logger.IsEnabled(LogLevel.Information))
+            {
+                _logMessages.Response(requestId, response.Status, response.ReasonPhrase, FormatHeaders(response.Headers), seconds);
+            }
+        }
+        else
+        {
+            if (ClientModelEventSource.Log.IsEnabled(EventLevel.Informational, EventKeywords.None))
```

Comment:
This pattern scares me a bit.  I guarantee that we'll forget to update the level here if we tweak it in the event source.   I don't have a good thought for how to guard against that other than removing the level here.

Maybe we can just add a giant comment in the event source and that'll be good enough since they don't often change?

Reply:
Doing the checks here limits some of the code duplication with formatting and sanitizing. But this is a really good point about the log /event level. If I move the formatting/sanitization back to the ClientModelLogMessages/ClientModelEventSource files the benefit of not forgetting to update the levels may outweigh the minimal code duplication.

Reply:
Any thought on creating a small consts class to hold the ids?  Something like:

```csharp
internal class LogEventIds
{
    public const SomeLog = 1;
    public const OtherLog = 2;
}
```

That would allow them to be shared between the different uses, I believe.

Reply:
Fair point.   The giant comment may be the best guard that we can do without causing other problems.